normalizer_lib.py

```diff
@@ -1,205 +1,218 @@
-
-    Returns:
-        The line or the line replaced with <SENTENCE_REJECTED>.
-    """
-    valid_token = INITIAL_PUNCTUATION.ques + Verbalizable.verbalizable() + FINAL_PUNCTUATION.ques
-    remove_extra_whitespace = (string @ REMOVE_EXTRA_WHITESPACE).string()
-    split_string = remove_extra_whitespace.split(" ")
-    for token in split_string:
-        if difference(acceptor(token), valid_token).num_states() != 0:
-            return "<SENTENCE_REJECTED>"
-    return string
-
-
-# Detach punctuation from words
-# e.g. "Who are you?" -> "Who are you ?"
-
-NON_GRAPHEME_PUNCT = difference(PUNCTUATION, GRAPHEMES)
-
-SEPARATE_PUNCTUATION = (
-    cdrewrite(
-        pynutil.insert(" "),
-        union("[BOS]", byte.SPACE) + union(NON_GRAPHEME_PUNCT),
-        (union(PUNCTUATION, Verbalizable.verbalizable()) +
-         union("[EOS]", byte.SPACE, Verbalizable.verbalizable())),
-        SIGMA_STAR) @
-    cdrewrite(
-        pynutil.insert(" "),
-        union(Verbalizable.verbalizable(), PUNCTUATION),
-        PUNCTUATION + union("[EOS]", byte.SPACE, PUNCTUATION),
-        SIGMA_STAR))
-
-
-# Delete freestanding punctuation
-# e.g. "hi there ?" -> "hi there
-
-DELETE_FREESTANDING_PUNCTUATION = cdrewrite(
-    pynutil.delete(PUNCTUATION),
-    union("[BOS]", byte.SPACE),
-    union("[EOS]", byte.SPACE),
-    SIGMA_STAR)
-
-
-def apply_fst_rules(string: str) -> str:
-    """Applies FST rewrite rules."""
-    return string if (string == "<STRING_REJECTED>") else (
-        string
-        @ LANGUAGE_SPECIFIC_NORM
-        @ SEPARATE_PUNCTUATION
-        @ DELETE_FREESTANDING_PUNCTUATION
-        @ REMOVE_EXTRA_WHITESPACE
-        ).optimize().string().rstrip()
-
-
-def token_normalizer(string: str) -> str:
-    """Normalizes text by applying FST rewrite rules.
-
-    Args:
-        string: A line from a corpus.
-
-    Returns:
-        The normalized line from the corpus.
-    """
-    string = unicodedata.normalize("NFC", string.lower())
-    filtered_string = pass_only_valid_tokens(string)
-    return apply_fst_rules(filtered_string)
-
-
-def sentence_normalizer(string: str) -> str:
-    """Normalizes text by applying FST rewrite rules.
-
-    Args:
-        string: A line from a corpus.
-
-    Returns:
-        The normalized line from the corpus.
-    """
-    string = unicodedata.normalize("NFC", string.lower())
-    filtered_string = pass_only_valid_sentences(string)
-    return apply_fst_rules(filtered_string)
+            filter_string = self.pass_only_valid_tokens(
+                unicode_normalize)
+        return filter_string
+
+
+    def token_normalizer(self, string: str) -> str:
+        """Normalizes text by applying FST rewrite rules.
+
+        Args:
+            string: A line from a corpus.
+
+        Returns:
+            The normalized line from the corpus.
+        """
+        return self.apply_fst_rules(self.preprocess_string(string, "token")
+                                    ).strip()
```

Comment:
Put the strip() call in the helper too.

Reply:
Moving it there didn't actually work, since it's more spaces might still exist after the `remove_extra_whitespace` step. However, I changed the right context of that deletion from `byte.SPACE` to `union(byte.SPACE, "[EOS]")`, so it will now strip all extra whitespace at the end of a line.